### ISSUE TITLE
Switch `systemd.user.startServices` default to `sd-switch`

### ISSUE THREAD

**Issue by rycee** [nix-community/home-manager#5452]
### Description

With version 0.4.0 of [sd-switch](https://git.sr.ht/~rycee/sd-switch), it should be more robust and work out of the box for most people. We should therefore switch to using it by default.

Sd-switch is a tool that automatically starts/stops/reloads systemd services when activating a Home Manager configuration.

You can use it today by adding `systemd.user.startServices = "sd-switch"` to your configuration. If it works well for you then consider adding a thumbs up reaction to this issue. If it does not work OK, then please add a comment outlining the issue you have. Make sure you have an up-to-date version of Nixpkgs unstable.

**Comment by drupol** [comment]
Enabled, so far so good on my side.

**Comment by rycee** [comment]
@drupol Thanks, that's good 🙂

**Comment by Janik-Haag** [comment]
@rycee I have noticed a small behavior change since switching to `sd-switch`. When I stop a user unit declared with home-manager, and then do a `nixos-rebuild switch` the unit starts again which didn't happen before.

**Comment by rycee** [comment]
@Janik-Haag Thanks for the feedback. This is probably an effect of sd-switch not starting new services directly, rather it starts all active targets. Doing it that way makes systemd figure out which of the new service units should be started and starts them.

Starting the target also starts stopped services, which is what you see. Shouldn't be terribly tricky to fix. I'll make sd-switch do the figuring out and start parts instead.

**Comment by rycee** [comment]
@Janik-Haag I've released a new version of sd-switch that is a bit more restrictive about starting unmanaged services. It will still start up managed services that are wanted by an active target, though. I'm not certain there is much that can be done about that since I don't know of a good way to detect "manually stopped" from "not running for other reasons". The new version should be available in all channels now.

**Comment by CyberShadow** [comment]
> I don't know of a good way to detect "manually stopped" from "not running for other reasons".

That sounds like an interesting and tricky problem, can you share more details?

Would it make sense for `sd-switch` to use the following logic:
1. If a unit was enabled in the previous configuration, but disabled/absent in the new configuration, then stop it
2. If a unit was disabled/absent in the previous configuration, but enabled in the new configuration, then start it

By the way, I noticed that `sd-switch` is described as a "third party application". But isn't it technically first-party, since it was created by the author of home-manager? :slightly_smiling_face:

**Comment by jtrees** [comment]
My use case is simple but using `sd-switch` works perfectly for it. The default method failed to restart the service when my config changed.

Here's my config snippet:

```nix
systemd.user.services.hyprpaper = {
  Unit = {
    Description = "desktop wallpaper";
    X-Restart-Triggers = [ "${config.home-manager.users.foo.xdg.configFile.hyprpaper.source}" ];
  };
  Install = {
    WantedBy = [ "graphical-session.target" ];
  };
  Service = {
    ExecStart = lib.getExe pkgs.hyprpaper;
  };
};
```

Thanks!

**Comment by Shados** [comment]
@rycee sd-switch does not appear to correctly handle services with `RefuseManualStart = true` and will attempt to manually start a service I have using that:
```
Error: Error switching

Caused by:
    0: Failed to perform post-reload tasks
    1: Failed to start unit power-switch-unlock.service
    2: org.freedesktop.systemd1.OnlyByDependency: Operation refused, unit power-switch-unlock.service may be requested by dependency only (it is configured to refuse manual start/stop).
```

The legacy switcher works fine for this, would be good to get that fixed before it's deprecated entirely...

**Comment by jackmac92** [comment]
I've been using sd-switch for a while, but since upgrading to 24.11 I keep getting

```
...
Activating reloadSystemd
Starting units: fusuma.service, go-chromecast-to-sqlite.service, goss-status-check.service, kotaemon.service, nfs-rclone-sync.target, prime-memory.service, sync-all-mailboxes.target, tray.target
Error: Error switching

Caused by:
    0: Failed to perform post-reload tasks
    1: timed out waiting on channel
```

which I'm not sure how to debug. I don't have any failing user services

**Comment by rycee** [comment]
@Shados Could you give it another try now. I released a new version of sd-switch a few days ago that may behave better when it comes to RefuseManual{Start,Stop}.

**Comment by Shados** [comment]
@rycee no dice; I actually happened to update things yesterday, and am now using sd-switch 0.5.2, but:
```
Activating reloadSystemd
Starting units: hm-graphical-session.target, openbox-graphical-session.target, power-switch-unlock.service, tray.target
Error: Error switching

Caused by:
    0: Failed to perform post-reload tasks
    1: Failed to start unit power-switch-unlock.service
    2: org.freedesktop.systemd1.OnlyByDependency: Operation refused, unit power-switch-unlock.service may be requested by dependency only (it is configured to refuse manual start/stop).
```

[Looking at sd-switch's code](https://git.sr.ht/~rycee/sd-switch/tree/42f2196de3f02728154ba01f09e7a6f30e4335a9/item/src/systemd/systemctl.rs#L272): you're checking if `RefuseManualStart/Stop` are equal to `yes`, but I have them set to `true` (as that's how they're specified in `man systemd.unit`). More broadly, `man systemd.syntax` shows that boolean values in systemd unit files (and other configuration files) have a pretty broad range of possible representations of boolean values you'll need to account for:
```
       Boolean arguments used in configuration files can be written in
       various formats. For positive settings the strings 1, yes, true
       and on are equivalent. For negative settings, the strings 0, no,
       false and off are equivalent.
```

It'd probably be a good idea to read through that man page in its entirety if you're going to parse unit files. Might be worthwhile writing a Serde crate for systemd configuration files, assuming there isn't already an acceptable one.

-----


Not directly related, but I'm also not sure why sd-switch is trying to start those targets, as they're already active and have been since this graphical session started:
```
$ systemctl --user status hm-graphical-session.target openbox-graphical-session.target tray.target
● hm-graphical-session.target - Home Manager X session
     Loaded: loaded (/home/shados/.config/systemd/user/hm-graphical-session.target; linked; preset: ignored)
     Active: active since Wed 2024-12-18 04:38:36 AEDT; 9h ago
 Invocation: b91e68ed138c43af828cd3e9e56c990b

Dec 18 04:38:36 dreamlogic systemd[4753]: Reached target Home Manager X session.

● openbox-graphical-session.target - Openbox Xorg session
     Loaded: loaded (/home/shados/.config/systemd/user/openbox-graphical-session.target; linked; preset: ignored)
     Active: active since Wed 2024-12-18 04:38:36 AEDT; 9h ago
 Invocation: dd95a85b8a114d42ba384242b101cbcb

Dec 18 04:38:36 dreamlogic systemd[4753]: Reached target Openbox Xorg session.

● tray.target - Home Manager System Tray
     Loaded: loaded (/home/shados/.config/systemd/user/tray.target; linked; preset: ignored)
     Active: active since Wed 2024-12-18 04:38:36 AEDT; 9h ago
 Invocation: 2d9e55dafc28484fb4f964f9a3d86908

Dec 18 04:38:36 dreamlogic systemd[4753]: Reached target Home Manager System Tray.
```

**Comment by rycee** [comment]
@Shados I just made a new release 0.5.3 that I think may resolve the isssue, apparently I missed one case where the unit start happens. If you want, then you can try patching Nixpkgs with https://github.com/NixOS/nixpkgs/pull/366860 and see if it helps. Otherwise it should be in unstable soonish.

About the yes/no, the output from `systemctl show` normalizes Boolean values to "yes" and "no", so it should be sufficient to look for those alternatives.

**Comment by Shados** [comment]
> I just made a new release 0.5.3 that I think may resolve the isssue, apparently I missed one case where the unit start happens. If you want, then you can try patching Nixpkgs with [NixOS/nixpkgs#366860](https://github.com/NixOS/nixpkgs/pull/366860) and see if it helps. Otherwise it should be in unstable soonish.
Yep, 0.5.3 did the trick, thanks! It is still trying to start the already-active targets as I mentioned previously, but at least that's a harmless no-op (although it does make for noisier HM activation output).

> About the yes/no, the output from `systemctl show` normalizes Boolean values to "yes" and "no", so it should be sufficient to look for those alternatives.

Ah, my bad.

**Comment by rycee** [comment]
@Shados Great, thanks for testing and confirming that if works! About the targets being started each time I have some memory of it being necessary in some case but it may be that with other improvements in sd-switch it now is unnecessary. I will investigate this a bit further when I have time.
